Dependencies/xr/Source/OpenXR/SceneUnderstanding.cpp

```diff
@@ -0,0 +1,497 @@
+
+        XrSpaceLocation viewInLocal{ XR_TYPE_SPACE_LOCATION };
+        CHECK_XRCMD(xrLocateSpace(m_viewSpace.Get(), space, time, &viewInLocal));
+
+        // TODO: support Frustum as needed
+        XrSceneSphereBoundMSFT sphere{};
+        XrSceneOrientedBoxBoundMSFT box{};
+        switch (m_boundaryType)
+        {
+        case DetectionBoundaryType::Sphere:
+        {
+            if (xr::math::Pose::IsPoseValid(viewInLocal)) {
+                sphere.center = viewInLocal.pose.position;
+                sphere.radius = m_sphereRadius;
+            }
+
+            computeInfo.bounds.sphereCount = 1;
+            computeInfo.bounds.spheres = &sphere;
+        }
+        break;
+        case DetectionBoundaryType::Box:
+        {
+            if (xr::math::Pose::IsPoseValid(viewInLocal)) {
+                box.pose = viewInLocal.pose;
+                box.extents = m_boxDimensions;
+            }
+            
+            computeInfo.bounds.boxCount = 1;
+            computeInfo.bounds.boxes = &box;
+        }
+        break;
+        }
+
+        CHECK_XRCMD(extensions.xrComputeNewSceneMSFT(m_sceneObserverHandle.Get(), &computeInfo));
+
+        m_lastSceneUpdate = time;
+        m_state = State::ComputingScene;
+    }
+
+    void HandleComputingScene(const XrTime time, const XrSpace& space, const XrSupportedExtensions& extensions)
+    {
+        XrSceneComputeStateMSFT state{};
+        CHECK_XRCMD(extensions.xrGetSceneComputeStateMSFT(m_sceneObserverHandle.Get(), &state));
+        if (state == XR_SCENE_COMPUTE_STATE_COMPLETED_MSFT) {
+            m_sceneHandle = xr::CreateScene(extensions, m_sceneObserverHandle.Get());
+
+            const auto sceneObjects = xr::GetSceneObjects(extensions, m_sceneHandle.Get());
+
+            std::vector<XrSceneObjectKeyMSFT> sceneObjectKeys;
+            sceneObjectKeys.resize(sceneObjects.size());
+            for (size_t i = 0; i < sceneObjectKeys.size(); i++)
+            {
+                sceneObjectKeys[i] = sceneObjects.at(i).sceneObjectKey;
+            }
+
+            XrSceneObjectsLocateInfoMSFT locateInfo{ XR_TYPE_SCENE_OBJECTS_LOCATE_INFO_MSFT };
+            locateInfo.baseSpace = space;
+            locateInfo.time = time;
+            locateInfo.sceneObjectCount = static_cast<uint32_t>(sceneObjects.size());
+            locateInfo.sceneObjectKeys = sceneObjectKeys.data();
+
+            m_sceneObjectLocations.resize(sceneObjects.size());
+            XrSceneObjectLocationsMSFT locations{ XR_TYPE_SCENE_OBJECT_LOCATIONS_MSFT };
+            locations.sceneObjectCount = static_cast<uint32_t>(m_sceneObjectLocations.size());
+            locations.sceneObjectLocations = m_sceneObjectLocations.data();
+
+            CHECK_XRCMD(extensions.xrLocateSceneObjectsMSFT(m_sceneHandle.Get(), &locateInfo, &locations));
+
+            // Get a set of the known objects to understand if any objects were removed
+            std::set<XrSceneObjectKeyMSFT> knownObjects{};

```

Comment:
nit: consider `std::unordered_set`. Might be somewhat more efficient for this use case.

Reply:
sounds good, I didn't know you could iterate over unordered_sets